src/main/java/org/jetbrains/bsp/bazel/server/BepServer.java

```diff
@@ -88,156 +86,146 @@ public void publishLifecycleEvent(
+    if (event.getId().hasNamedSet()) {
+      namedSetsOfFiles.put(event.getId().getNamedSet().getId(), event.getNamedSetOfFiles());
+    }
   }
 
-  private void processStdErrDiagnostics(BuildEventStreamProtos.Progress progress) {
-    HashMap<String, List<Diagnostic>> fileDiagnostics = new HashMap<>();
-    Arrays.stream(progress.getStderr().split("\n"))
-        .filter(
-            error ->
-                error.contains("ERROR")
-                    && (error.contains("/" + workspace + ":") || error.contains("/" + build + ":")))
-        .forEach(
-            error -> {
-              String erroredFile = error.contains(workspace) ? workspace : build;
-              String[] lineLocation;
-              String fileLocation;
+  private void processBuildStartedEventIfPresent(BuildEventStreamProtos.BuildEvent event) {
+    if (event.hasStarted() && event.getStarted().getCommand().equals("build")) {
+      processedStartedEvent(event.getStarted());
+    }
+  }
 
-              if (error.contains(" at /")) {
-                int endOfMessage = error.indexOf(" at /");
-                String fileInfo = error.substring(endOfMessage + 4);
-                int urlEnd = fileInfo.indexOf(erroredFile) + erroredFile.length();
-                fileLocation = fileInfo.substring(0, urlEnd);
-                lineLocation = fileInfo.substring(urlEnd + 1).split(":");
-              } else {
-                int urlEnd = error.indexOf(erroredFile) + erroredFile.length();
-                fileLocation = error.substring(error.indexOf("ERROR: ") + 7, urlEnd);
-                lineLocation = error.substring(urlEnd + 1).split("(:)|( )");
-              }
-              System.out.println("Error: " + error);
-              System.out.println("File location: " + fileLocation);
-              System.out.println("Line location: " + Arrays.toString(lineLocation));
-              Position position =
-                  new Position(
-                      Integer.parseInt(lineLocation[0]), Integer.parseInt(lineLocation[1]));
-              Diagnostic diagnostic = new Diagnostic(new Range(position, position), error);
-              diagnostic.setSeverity(DiagnosticSeverity.ERROR);
-              List<Diagnostic> diagnostics =
-                  fileDiagnostics.getOrDefault(fileLocation, new ArrayList<>());
-              diagnostics.add(diagnostic);
-              fileDiagnostics.put(fileLocation, diagnostics);
-            });
+  private void processFinishedEventIfPresent(BuildEventStreamProtos.BuildEvent event) {
+    if (event.hasFinished()) {
+      processFinishedEvent(event.getFinished());
+    }
+  }
+
+  private void processCompletedEventIfPresent(BuildEventStreamProtos.BuildEvent event)
+      throws IOException {
+    if (event.hasCompleted()) {
+      processCompletedEvent(event.getCompleted());
+    }
+  }
+
+  private void processActionEventIfPresent(BuildEventStreamProtos.BuildEvent event)
+      throws IOException {
```

Comment:
why does it throw?

Reply:
How would you do it otherwise? The try-catch in handleEvent aggregates and handles all possible exceptions from the first _parseFrom_ and event processors (which are essentially also parsing functions). If I were to wrap everything locally with try-catch in these processors, I'd have to have 5 of them (in _processActionEvent_ specifically _collectDiagnostics_ throw).